File: server/Netptune.Core/Entities/ProjectTask.cs
using System.Text.Json.Serialization;

using Netptune.Core.BaseEntities;
using Netptune.Core.Enums;
using Netptune.Core.Relationships;

namespace Netptune.Core.Entities;

public record ProjectTask : WorkspaceEntity<int>
{
    public string Name { get; set; } = null!;

    public string? Description { get; set; }

    public int StatusId { get; set; }

    public int ProjectScopeId { get; set; }

    public TaskPriority? Priority { get; set; }

    public EstimateType? EstimateType { get; set; }

    public decimal? EstimateValue { get; set; }

    public DateOnly? StartDate { get; set; }

    public DateOnly? DueDate { get; set; }

    public string? ExternalId { get; set; }

    #region ForeignKeys

    public int? ProjectId { get; set; }

    public int? SprintId { get; set; }

    #endregion

    #region NavigationProperties

    [JsonIgnore]
    public Status? Status { get; set; }

    [JsonIgnore]
    public ICollection<ProjectTaskAppUser> ProjectTaskAppUsers { get; set; } = new HashSet<ProjectTaskAppUser>();

    [JsonIgnore]
    public Project? Project { get; set; }

    [JsonIgnore]
    public Sprint? Sprint { get; set; }

    [JsonIgnore]
    public ICollection<ProjectTaskInBoardGroup> ProjectTaskInBoardGroups { get; set; } = new HashSet<ProjectTaskInBoardGroup>();

    [JsonIgnore]
    public ICollection<ProjectTaskTag> ProjectTaskTags { get; set; } = new HashSet<ProjectTaskTag>();

    [JsonIgnore]
    public ICollection<TaskFile> Files { get; set; } = new HashSet<TaskFile>();

    [JsonIgnore]
    public ICollection<Tag> Tags { get; set; } = new HashSet<Tag>();

    #endregion
}

File: client/src/app/core/models/project-task.ts
import { EstimateType } from '../enums/estimate-type';
import { TaskPriority } from '../enums/task-priority';
import { SprintStatus } from '../enums/sprint-status';
import { StatusCategory } from './status';
import { AppUser } from './appuser';
import { Basemodel } from './basemodel';
import { Project } from './project';
import { Workspace } from './workspace';

export interface ProjectTask extends Basemodel {
  name: string;
  description: string;
  statusId: number;
  statusName: string;
  statusKey: string;
  statusColor?: string | null;
  statusCategory: StatusCategory;
  priority: TaskPriority | null;
  estimateType: EstimateType | null;
  estimateValue: number | null;
  dueDate?: string | null;

  sortOrder: number;

  project: Project;
  projectId: number;
  projectScopeId: number;

  sprintId?: number | null;
  sprintName?: string | null;
  sprintStatus?: SprintStatus | null;

  workspace: Workspace;
  workspaceId: number;

  owner: AppUser;
  ownerId: string;
}

export interface AddProjectTaskRequest {
  name: string;
  description?: string;
  statusId?: number;

  projectId: number;
  boardGroupId?: number;
  sprintId?: number | null;

  sortOrder?: number;

  assigneeId?: string;

  priority?: TaskPriority;
  estimateType?: EstimateType;
  estimateValue?: number;
  dueDate?: string | null;
}
